core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -238,7 +241,13 @@ private[spark] class TaskSchedulerImpl(
   private[scheduler] def createTaskSetManager(
       taskSet: TaskSet,
       maxTaskFailures: Int): TaskSetManager = {
-    new TaskSetManager(this, taskSet, maxTaskFailures, blacklistTrackerOpt)
+    val finishedPartitions =
+      stageIdToFinishedPartitions.getOrElseUpdate(taskSet.stageId, new HashSet[Int])
+    // filter the task which has been finished by previous attempts
+    val tasks = taskSet.tasks.filterNot{ t => finishedPartitions(t.partitionId) }
```

Comment:
I'd change this comment a bit, to explain why this is necessary, even briefly.  Something like
"The DAGScheduler does not always know all the tasks that have been completed by other tasksets when completing a stage, so we do an extra filter here, while holding the TaskSchedulerImpl lock.  See SPARK-25250 and markPartitionCompletedInAllTaskSets()`